clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/agents/TitusStreamingUpdateAgent.java

```
@@ -185,37 +187,73 @@ public void executeAgent(Agent agent) {
-              updateJob(notification.getJobUpdate().getJob());
-              break;
-            case TASKUPDATE:
-              updateTask(notification.getTaskUpdate().getTask());
-              break;
-            case SNAPSHOTEND:
-              lastUpdate.set(0); // last update should already be initialized to 0 but this forces an update regardless
-              log.info("Snapshot finished {}", getAgentType());
-              tasks.keySet().retainAll(jobs.keySet());
-              break;
+
+        while (continueStreaming(startTime)) {
+          try {
+            while (notificationIt.hasNext() && continueStreaming(startTime)) {
+              JobChangeNotification notification = notificationIt.next();
+              switch (notification.getNotificationCase()) {
+                case JOBUPDATE:
+                  updateJob(notification.getJobUpdate().getJob());
+                  break;
+                case TASKUPDATE:
+                  updateTask(notification.getTaskUpdate().getTask(), snapshotComplete);
+                  break;
+                case SNAPSHOTEND:
+                  lastUpdate.set(0); // last update should already be initialized to 0 but this forces it
+                  log.info("{} snapshot finished in {}ms", getAgentType(), System.currentTimeMillis() - startTime);
+                  tasks.keySet().retainAll(jobs.keySet());
+                  if (snapshotComplete) {
+                    log.error("{} received >1 SNAPSHOTEND events, this is unexpected and may be handled incorrectly",
+                      getAgentType()
+                    );
+                  }
+                  snapshotComplete = true;
+                  break;
+              }
+
+              if (snapshotComplete) {
+                writeToCache(!savedSnapshot);
+                if (!savedSnapshot) {
+                  savedSnapshot = true;
+                }
+              }
+            }
+          } catch (io.grpc.StatusRuntimeException e) {
```

Streaming agent invocations were consistently hitting gRPC `DEADLINE_EXCEEDED` errors and terminating early but the exception was getting swallowed within this future.